content/blog/2022-05-19-A_Type_and_Effect_System_for_Deterministic_Parallel_Java/index.md

```diff
@@ -0,0 +1,123 @@
++++
+title = "A Type and Effect System for Deterministic Parallel Java"
+[extra]
+bio = """
+  [Nikita Lazarev](https://www.nikita.tech/) is an PhD student at Cornell's ECE department;
+  Susan Garry is an undergraduate student at Cornell's CS department.
+"""
+[[extra.authors]]
+name = "Nikita Lazarev"
+link = "https://www.nikita.tech/"  # Links are optional.
+[[extra.authors]]
+name = " Susan Garry"
++++
+
+### Paper review
+
+["A Type and Effect System for Deterministic Parallel Java", Robert L. Bocchino, Vikram S. Adve, Danny Dig, Sarita V. Adve, Stephen Heumann, Rakesh Komuravelli, Jeffrey Overbey, Patrick Simmons, Hyojin Sung, and Mohsen Vakilian. OOPSLA 2009](http://dpj.cs.illinois.edu/DPJ/Publications_files/DPJ-OOPSLA-2009.pdf)
+
+### Introduction
+
+Multithreading is notoriously difficult to reason about and error-prone. This happens due to potential race conditions that can be arise when two threads try to access the same region in memory simultaneously. Traditional approaches using locks (e.g. mutexes) can cause a deadlock and are also difficult to reason about and debug.
+
+In this paper, the authors propose making concurrency in Java safe and lock-free by using a type and effect system that involves partitioning and annotating Java programs. They introduce novel mechanisms that expand on the capabilities of previous type and effect systems, allowing for concurrent operations on recursive data structures via **region nesting**, concurrent array operations via **index-parameterized arrays** and **subarrays**, and concurrent writes to the same location in memory via **commutativity annotations**. They develop a new extension for Java called DPJ (Deterministic Parallel Java) which additionally comes with the formal verification functionality. The paper demonstrates how DPJ can be used to implement realistic algorithms and verify their thread-safeness.
+
+### Basic Idea
+
+The paper introduces a set of annotations for unsafe memory accesses. The systems defines two main classes of annotations:
+* **regions** - the programmer names portions of the heap and assigns each object and variable to a region of the heap;
+* **effect annotations** - denotes what regions a method reads and writes from, e.g. ```reads <region-list> writes <region-list>```, where ```<region-list>``` is a list of regions as defined above.
+
+The code below shows an example of a class definition (a Tree class for a tree data structure) with annotations. The example is taken from the original paper, and we added more comments in it to show the functionality.
```

Comment:
Maybe one or two sentences _after_ the example could really help clarify the high-level intuition and takeaways that the reader is supposed to get from reading the example?